### ISSUE TITLE
water

### ISSUE THREAD

**Issue by Aiswaryavijayy** [Aiswaryavijayy/water-quality-anaysis-using-explainable-ai#2]
`import numpy as np
import pandas as pd
import shap
import matplotlib.pyplot as plt
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import RobustScaler
from sklearn.utils.class_weight import compute_sample_weight

# Define features globally
features = ['Temperature', 'Turbidity', 'pH', 'Electrical Conductivity', 'Dissolved Oxygen']

np.random.seed(42)
n_suitable = 500
n_unsuitable = 500

df_suitable = pd.DataFrame({
    'pH': np.random.uniform(6.8, 8.2, n_suitable),
    'Temperature': np.random.uniform(12, 28, n_suitable),
    'Turbidity': np.random.uniform(1, 40, n_suitable),
    'Dissolved Oxygen': np.random.uniform(6, 9, n_suitable),
    'Electrical Conductivity': np.random.uniform(300, 1000, n_suitable)
})
df_suitable['Suitability'] = 1

pH_unsuitable = np.where(np.random.rand(n_unsuitable) < 0.5,
                         np.random.uniform(6.0, 6.4, n_unsuitable),
                         np.random.uniform(8.6, 9.0, n_unsuitable))
Temperature_unsuitable = np.where(np.random.rand(n_unsuitable) < 0.5,
                                  np.random.uniform(5, 9, n_unsuitable),
                                  np.random.uniform(31, 35, n_unsuitable))
Turbidity_unsuitable = np.random.uniform(55, 60, n_unsuitable)
Dissolved_Oxygen_unsuitable = np.where(np.random.rand(n_unsuitable) < 0.5,
                                       np.random.uniform(4, 4.9, n_unsuitable),
                                       np.random.uniform(10.1, 12, n_unsuitable))
Electrical_Conductivity_unsuitable = np.where(np.random.rand(n_unsuitable) < 0.5,
                                              np.random.uniform(50, 90, n_unsuitable),
                                              np.random.uniform(1510, 1600, n_unsuitable))

df_unsuitable = pd.DataFrame({
    'pH': pH_unsuitable,
    'Temperature': Temperature_unsuitable,
    'Turbidity': Turbidity_unsuitable,
    'Dissolved Oxygen': Dissolved_Oxygen_unsuitable,
    'Electrical Conductivity': Electrical_Conductivity_unsuitable
})
df_unsuitable['Suitability'] = 0

df = pd.concat([df_suitable, df_unsuitable], ignore_index=True)
df = df.sample(frac=1, random_state=42).reset_index(drop=True)

n_noise = int(0.05 * len(df))
noise_indices = np.random.choice(df.index, size=n_noise, replace=False)
df.loc[noise_indices, 'Suitability'] = 1 - df.loc[noise_indices, 'Suitability']

X = df[features]
y = df['Suitability']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = RobustScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

sample_weight = compute_sample_weight(class_weight='balanced', y=y_train)

model = XGBClassifier(n_estimators=500, max_depth=5, learning_rate=0.05, random_state=42, eval_metric='logloss')
model.fit(X_train_scaled, y_train, sample_weight=sample_weight)

y_pred = model.predict(X_test_scaled)
accuracy_val = accuracy_score(y_test, y_pred)
print("Test Accuracy: {:.2f}%".format(accuracy_val * 100))

explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_test_scaled)

shap.summary_plot(shap_values, X_test, feature_names=features)

def predict_unseen():
    print("\nEnter values for a new water quality sample:")
    input_data = {}
    for feat in features:
        input_data[feat] = float(input(f"Enter {feat}: "))
    df_input = pd.DataFrame([input_data])
    scaled_input = scaler.transform(df_input)

    probas = model.predict_proba(scaled_input)[0]

    alerts = []
    if not (6.5 <= input_data['pH'] <= 8.5):
        alerts.append("pH")
    if not (10 <= input_data['Temperature'] <= 30):
        alerts.append("Temperature")
    if input_data['Turbidity'] > 5:
        alerts.append("Turbidity")
    if not (5 <= input_data['Dissolved Oxygen'] <= 10):
        alerts.append("Dissolved Oxygen")
    if not (100 <= input_data['Electrical Conductivity'] <= 1500):
        alerts.append("Electrical Conductivity")

    pred = model.predict(scaled_input)[0]
    confidence = probas[pred]

    shap_vals_unseen = explainer.shap_values(scaled_input)

    for param in alerts:
        if param == 'pH':
            print("\u274C pH out of range (6.5-8.5)-- Unsafe pH levels can make water acidic or alkaline, affecting health.\n   Proposed Treatment Approach - Consider adding lime to increase pH or using acid to lower it.")
        elif param == 'Temperature':
            print("\u274C Temperature out of range (10-30°C) - Extreme temperatures can affect water chemistry and microbial activity.\n   Proposed Treatment Approach - Use cooling or aeration methods to regulate temperature.")
        elif param == 'Turbidity':
            print("\u274C Turbidity above acceptable limit (0-5 NTU) - High turbidity indicates excessive particles in water, reducing clarity.\n   Proposed Treatment Approach - Use filtration or coagulation techniques.")
        elif param == 'Dissolved Oxygen':
            print("\u274C Dissolved Oxygen out of range (5-10 mg/L) - Too low oxygen harms aquatic life, while excessive oxygen can indicate contamination.\n   Proposed Treatment Approach - Aeration or adding oxygen can improve levels.")
        elif param == 'Electrical Conductivity':
            print("\u274C Electrical Conductivity out of range (100-1500 µS/cm) - High values indicate possible contamination with salts or heavy metals.\n   Proposed Treatment Approach - Reverse osmosis or ion exchange can help reduce conductivity.")

    if not alerts:
        print("\n✅ Water is predicted to be SUITABLE for drinking.")
    else:
        print("\n❌ Water is predicted to be NOT SUITABLE for drinking.")

    print("Confidence: {:.2f}%".format(confidence * 100))

    print("\nExplanation of Prediction:")
    for i, feat in enumerate(features):
        impact = shap_vals_unseen[0][i]
        reason = "(Outside safe range - Affects water quality)" if feat in alerts else "(Within safe range)"
        print(f"  - {feat}: {impact:.4f} {reason}")

predict_unseen()`
